pkg/webhook/resources/virtualmachine/mutator.go

```diff
@@ -454,3 +449,19 @@ func hostDevicesPresent(vm *kubevirtv1.VirtualMachine) bool {
 
 	return false
 }
+
+func hostDevicesOvercommitNeeded(oldVM, newVM *kubevirtv1.VirtualMachine) bool {
+	// if host devices or GPUs are added or removed then resource update is needed
+	if len(newVM.Spec.Template.Spec.Domain.Devices.HostDevices) != len(oldVM.Spec.Template.Spec.Domain.Devices.HostDevices) || len(newVM.Spec.Template.Spec.Domain.Devices.GPUs) != len(oldVM.Spec.Template.Spec.Domain.Devices.GPUs) {
+		return true
+	}
+
+	// during upgrade path VMs with host devices are turned off. Post upgrade the memory needs to be patched
+	// to ensure devices with hostDevices/GPUs can be booted
+	if len(newVM.Spec.Template.Spec.Domain.Devices.HostDevices) > 0 || len(newVM.Spec.Template.Spec.Domain.Devices.GPUs) > 0 {
+		if newVM.Spec.Template.Spec.Domain.Memory != nil && (newVM.Spec.Template.Spec.Domain.Memory.Guest.Value() != newVM.Spec.Template.Spec.Domain.Resources.Requests.Memory().Value()) {
```

Comment:
Question:

The `newVM.Spec.Template.Spec.Domain.Memory.Guest.Value()` seems always differ with `newVM.Spec.Template.Spec.Domain.Resources.Requests.Memory().Value()`,  

will the above code always run, or the `newVM.Spec.Template.Spec.Domain.Memory.Guest.Value()` will be equal to `newVM.Spec.Template.Spec.Domain.Resources.Requests.Memory().Value()` after patch ? 

```

            hostDevices:
            - deviceName: nvidia.com/TU104GL_TESLA_T4
              name: pci-test-01-000008000
              
              
          machine:
            type: q35
          memory:
            guest: 8092Mi
          resources:
            limits:
              cpu: "4"
              memory: 8Gi
            requests:
              cpu: 250m
              memory: 5461Mi
        evictionStrategy: LiveMigrate
        hostname: focal-cuda-01
 ...
  status:
    conditions:
    - lastProbeTime: "2024-05-13T20:29:10Z"
      lastTransitionTime: "2024-05-13T20:29:10Z"
      message: Guest VM is not reported as running
      reason: GuestNotRunning
      status: "False"
      type: Ready
    - lastProbeTime: "null"
      lastTransitionTime: "null"
      message: VMI uses a PCI host devices
      reason: HostDeviceNotLiveMigratable
      status: "False"
      type: LiveMigratable
    - lastProbeTime: "null"
      lastTransitionTime: "2024-05-13T20:29:27Z"
      message: 'server error. command SyncVMI failed: "LibvirtError(Code=1, Domain=10,
        Message=''internal error: Process exited prior to exec: libvirt:  error :
        cannot limit locked memory of process 81 to 9558818816: Operation not permitted'')"null'
      reason: Synchronizing with the Domain failed.
      status: "False"
````

Reply:
the check will only be applied when hostdevices/gpus are present in the vm definition. When vm meets this condition the resource limits will be recalculated. This is already being done in v1.3.0 and v1.2.2-rc when a new vm is launched or an existing one is updated to add/remove hostdevices/gpus.
The checks addresses the upgrade path from 1.2.1 for VM's with host devices which were turned off during the upgrade. Powering on the vm is missed by the check currently as there is no change in number of hostdevices/gpus or vm resource requirements and we end up with the VM being unable to start and the error we see above.

Reply:
in addition to above, this check will be run on each VM stop/start. This will trigger a calculation of memory overcommit and a patch will be generated if needed. Thanks @w13915984028 for pointing out.